### PR TITLE
Peer honesty implementation

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added peer honesty implementation able to black list a public key if it insists to send bad messages. This is done currently only on consensus topic, other topics will follow.
- consensus now do not count messages in antiflood instance as it does not require to do so due to the improved filtering mechanism and peer honesty tracking process

### CODE REVIEW

File: cmd/node/config/ratings.toml [hunk 1]
```diff
@@ -40,3 +40,13 @@
...
     ValidatorDecreaseFactor = -4.0
     ConsecutiveMissedBlocksPenalty = 1.10
+
+[PeerHonesty]
+    #this value will be multiplied with the current value for a public key each DecayUpdateIntervalInSeconds seconds
+    #for the current settings, a pk will reach to value 0 after maximum 2h (if it stopped misbehaving or stopped sending good messages)
+    DecayCoefficient             = 0.9779
+    DecayUpdateIntervalInSeconds = 10
+    MaxScore                     = 100
+    MinScore                	 = -100
```
Comment (sasurobert) [inline cmd/node/config/ratings.toml:50]: indent

File: consensus/spos/constants.go [hunk 2]
```diff
@@ -5,16 +5,16 @@ const MaxThresholdPercent = 95
 
 // LeaderPeerHonestyIncreaseFactor specifies the factor with which the honesty of the leader should be increased
 // if it proposed a block or sent the final info, in its correct allocated slot/time-frame/round
-const LeaderPeerHonestyIncreaseFactor = 2.0
+const LeaderPeerHonestyIncreaseFactor = 2
 
 // ValidatorPeerHonestyIncreaseFactor specifies the factor with which the honesty of the validator should be increased
 // if it sent the signature, in its correct allocated slot/time-frame/round
-const ValidatorPeerHonestyIncreaseFactor = 1.0
+const ValidatorPeerHonestyIncreaseFactor = 1
```
Comment (sasurobert) [inline consensus/spos/constants.go:12]: maybe move these to config.

File: process/errors.go [hunk 3]
```diff
@@ -808,3 +808,24 @@ var ErrInvalidUserNameLength = errors.New("invalid user name length")
...
+var ErrInvalidMinScore = errors.New("invalid minimum score")
+
+// ErrInvalidMaxScore signals that an invalid maximum score was provided
+var ErrInvalidMaxScore = errors.New("invalid maximum score")
+
+// ErrInvalidUnitValue signals that an invalid unit value was provided
+var ErrInvalidUnitValue = errors.New("invalid unit value")
+
+// ErrInvalidBadPeerThreshold signals that an invalid bad peer threshold has been provided
+var ErrInvalidBadPeerThreshold = errors.New("invalid bad peer threshold")
```
Comment (sasurobert) [inline process/errors.go:831]: empty line to end of file

File: process/rating/peerHonesty/peerScore_test.go [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+		atomic.AddInt32(&numCalls, 1)
+	}
+	pph, _ := NewP2pPeerHonestyWithCustomExecuteDelayFunction(
+		cfg,
+		&mock.TimeCacheStub{},
+		&testscommon.CacherStub{},
+		handler,
+	)
+
+	time.Sleep(time.Second*3 + time.Millisecond*100) //this will call three times the handler
```
Comment (sasurobert) [inline process/rating/peerHonesty/peerScore_test.go:175]: will call the handler 3 times

File: process/rating/peerHonesty/peerScore.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+		log.Warn("p2pPeerHonesty.checkBlacklist",
+			"hex public key", hex.EncodeToString([]byte(ps.pk)),
+			"error", err)
+	}
+}
+
+// Close closes the running go routines related to this instance
+func (pph *p2pPeerHonesty) Close() error {
+	pph.cancelFunc()
+
```
Comment (sasurobert) [inline process/rating/peerHonesty/peerScore.go:273]: delete empty line

File: process/rating/peerHonesty/peerScore.go [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+	if change < 0 {
+		//TODO switch this to log.Trace in the future
+		log.Debug("p2pPeerHonesty.ChangeScore decrease",
+			"pk", hex.EncodeToString([]byte(pk)),
+			"current", fmt.Sprintf("%.2f", oldValue),
+			"change", fmt.Sprintf("%.2f", change),
+		)
+	}
+
+	ps.scores[topic] += change
```
Comment (sasurobert) [inline process/rating/peerHonesty/peerScore.go:202]: might use core.Max and core.Min here. No need to do that in checkBlackList

File: process/rating/peerHonesty/peerScore.go [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+		}
+
+		ps, ok := psObj.(*peerScore)
+		if !ok {
+			continue
+		}
+
+		for topic, score := range ps.scores {
+			score = score * pph.decayCoefficient
+			if score < approximateZero && score > 0 {
```
Comment (sasurobert) [inline process/rating/peerHonesty/peerScore.go:174]: add function check.IfZero(value, epsilon)

File: process/rating/peerHonesty/peerScore.go [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+	}
+}
+
+func (pph *p2pPeerHonesty) applyDecay() {
+	pph.mut.Lock()
+	defer pph.mut.Unlock()
+
+	keys := pph.cache.Keys()
+	for _, key := range keys {
+		psObj, _ := pph.cache.Get(key)
```
Comment (sasurobert) [inline process/rating/peerHonesty/peerScore.go:162]: why don't you treat the error instead of nil check ?

File: process/rating/peerHonesty/peerScore.go [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+		if score < pph.badPeerThreshold {
+			shouldBlacklist = true
+		}
+	}
+
+	if !shouldBlacklist {
+		return
+	}
+
+	if pph.blackListedPkCache.Has(ps.pk) {
```
Comment (andreibancioiu) [inline process/rating/peerHonesty/peerScore.go:258]: So we do not extend the `stay in the black list` period for already-blacklisted peers even if `shouldBlacklist == true`, right? Well, it makes sense since a blacklisted peer can't misbehave anymore anyway - at least not with effect (I am just thinking out loud).

File: process/rating/peerHonesty/peerScore_test.go [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/process/mock"
+	"github.com/ElrondNetwork/elrond-go/testscommon"
+	"github.com/stretchr/testify/assert"
+)
+
+// createMockPeerHonestyConfig creates a peer honesty config with reasonable values
+func createMockPeerHonestyConfig() config.PeerHonestyConfig {
+	return config.PeerHonestyConfig{
+		DecayCoefficient:             0.9995,
```
Comment (andreibancioiu) [inline process/rating/peerHonesty/peerScore_test.go:20]: I know I am exaggerating here, but can you use the same value as the one from `config.toml` perhaps?

File: process/rating/peerHonesty/peerScore_test.go [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+		cfg,
+		&mock.TimeCacheStub{},
+		&testscommon.CacherStub{},
+	)
+
+	assert.False(t, check.IfNil(pph))
+	assert.Nil(t, err)
+}
+
+func TestP2pPeerHonesty_Close(t *testing.T) {
```
Comment (andreibancioiu) [inline process/rating/peerHonesty/peerScore_test.go:159]: This test is not very short (in duration) - just a remark.

File: process/rating/peerHonesty/peerScore_test.go [hunk 12]
```diff
@@ -0,0 +1,418 @@
+package peerHonesty
```
Comment (andreibancioiu) [inline process/rating/peerHonesty/peerScore_test.go:1]: :+1:  Tests cover everything with the exception of impossible cases (bad casts etc.).

File: process/rating/peerHonesty/peerScore_test.go [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+
+	pks := []string{"pkMin", "pkMax", "pkNearZero", "pkZero", "pkValue"}
+	initial := []float64{cfg.MinScore, cfg.MaxScore, approximateZero / 2, 0, 28}
+	topic := "topic"
+
+	for idx := range pks {
+		pph.Put(pks[idx], topic, initial[idx])
+	}
+
+	pph.applyDecay()
```
Comment (andreibancioiu) [inline process/rating/peerHonesty/peerScore_test.go:406]: Perhaps add an extra test that calls `applyDecay` a lot of times, just to assert that all scores go to zero?

File: cmd/node/config/ratings.toml (hunk 1 above)
Comment (iulianpascalau) [reply cmd/node/config/ratings.toml:50]: done

File: process/errors.go (hunk 3 above)
Comment (iulianpascalau) [reply process/errors.go:831]: done

File: process/rating/peerHonesty/peerScore.go (hunk 5 above)
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:273]: done

File: process/rating/peerHonesty/peerScore.go (hunk 6 above)
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:202]: I can not use core max and core min as this will always rewrite to the minimum value. Moved the checks out of checkBlackList

File: process/rating/peerHonesty/peerScore.go (hunk 7 above)
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:174]: done

File: process/rating/peerHonesty/peerScore.go (hunk 8 above)
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:162]: done

File: consensus/mock/peerHonestyHandlerStub.go [hunk 14]
```diff
@@ -2,22 +2,14 @@ package mock
 
 // PeerHonestyHandlerStub -
 type PeerHonestyHandlerStub struct {
-	IncreaseCalled func(pk string, topic string, value float64)
-	DecreaseCalled func(pk string, topic string, value float64)
+	ChangeScoreCalled func(pk string, topic string, units int)
+	DecreaseCalled    func(pk string, topic string, value float64)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply consensus/mock/peerHonestyHandlerStub.go:6]: done

File: consensus/mock/peerHonestyHandlerStub.go [hunk 15]
```diff
@@ -2,22 +2,14 @@ package mock
...
-}
-
-// Decrease -
-func (phhs *PeerHonestyHandlerStub) Decrease(pk string, topic string, value float64) {
-	if phhs.DecreaseCalled != nil {
-		phhs.DecreaseCalled(pk, topic, value)
+// ChangeScore -
+func (phhs *PeerHonestyHandlerStub) ChangeScore(pk string, topic string, units int) {
+	if phhs.ChangeScoreCalled != nil {
+		phhs.ChangeScoreCalled(pk, topic, units)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply consensus/mock/peerHonestyHandlerStub.go:12]: done, remoed

File: process/errors.go [hunk 16]
```diff
@@ -808,3 +808,24 @@ var ErrInvalidUserNameLength = errors.New("invalid user name length")
 
 // ErrTxValueOutOfBounds signals that transaction value is out of bounds
 var ErrTxValueOutOfBounds = errors.New("tx value is out of bounds")
+
+// ErrNilBlackListedPkCache signals that a nil black listed public key cache
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/errors.go:812]: done

File: process/rating/peerHonesty/peerScore.go [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+
+	"github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+var log = logger.GetOrCreate("process/rating/peerhonesty")
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:19]: it just follows our pattern. I will keep this in order to comply with the rest of the project

File: process/rating/peerHonesty/peerScore.go [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+
+const float64Size = 8
+const defaultTopicSize = 20 //this is an approximate size, used for fast computing
+const approximateZero = 0.00001
+
+const minDecayCoefficient = 0.0
+const maxDecayCoefficient = 1.0
+const minDecayIntervalInSeconds = uint32(1)
+
+type peerScore struct {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:29]: yeah, great idea, will do just that

File: process/rating/peerHonesty/peerScore.go [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+	pk     string
+	scores map[string]float64
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:31]: done, renamed to scoresByTopic

File: process/rating/peerHonesty/peerScore.go [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+}
+
+func (ps *peerScore) size() int {
+	return len(ps.pk) + len(ps.scores)*(float64Size+defaultTopicSize)
+}
+
+func (ps *peerScore) String() string {
+	scores := make([]string, 0, len(ps.scores))
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:39]: added warning, shouldn't be the case of concurrent usage as the struct is unexported and also its constructor

File: process/rating/peerHonesty/peerScore.go [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+	minScore               float64
+	badPeerThreshold       float64
+	unitValue              float64
+	cache                  storage.Cacher
+	mut                    sync.RWMutex
+	blackListedPkCache     process.TimeCacher
+	cancelFunc             func()
+}
+
+// NewP2pPeerHonesty creates a new peer honesty handler able to manage a provided set of public keys
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:60]: added

File: process/rating/peerHonesty/peerScore.go [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	instance.cancelFunc = cancelFunc
+
+	go instance.executeDecay(ctx, instance.applyDecay)
+
+	return instance, nil
+}
+
+func checkParams(
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:90]: yeah, should ease the troubleshooting process

File: process/rating/peerHonesty/peerScore.go [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+	if !isBadPeerThresholdOk {
+		return fmt.Errorf("%w, BadPeerThreshold value should be in interval (MinScore, 0)",
+			process.ErrInvalidBadPeerThreshold,
+		)
+	}
+
+	return nil
+}
+
+func (pph *p2pPeerHonesty) executeDecay(ctx context.Context, handler func()) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:145]: done

File: process/rating/peerHonesty/peerScore.go [hunk 24]
```diff
@@ -0,0 +1,280 @@
...
+	ps := &peerScore{
+		scores: make(map[string]float64),
+		pk:     pk,
+	}
+	pph.cache.Put(key, ps, ps.size())
+
+	return ps
+}
+
+func (pph *p2pPeerHonesty) checkBlacklist(ps *peerScore) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:237]: added noLock suffix

File: process/rating/peerHonesty/peerScore.go (hunk 9 above)
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore.go:258]: exactly, since this is checked on each 10 seconds or so, it will prolong the blacklisted status beyond easy understanding of this mechanism

File: process/rating/peerHonesty/peerScore_test.go (hunk 10 above)
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore_test.go:20]: done

File: process/rating/peerHonesty/peerScore_test.go (hunk 11 above)
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore_test.go:159]: I know, shortened

File: process/rating/peerHonesty/peerScore_test.go (hunk 13 above)
Comment (iulianpascalau) [reply process/rating/peerHonesty/peerScore_test.go:406]: done
